[addons] add general addon callback functions to new style

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This add the general functions who used on several addons on the new addon
interface style.

In future all the basics like on Python addons can be added here.

Also is related documentation already there, but need improved in future (my bad english ;-) ).

<!--- Describe your change in detail -->

## Motivation and Context
Next step to change the addon interface. This way is later needed after all addons are changed.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h:
```diff
@@ -50,6 +50,11 @@
+#define ADDON_GLOBAL_VERSION_GENERAL                  "1.0.0"
+#define ADDON_GLOBAL_VERSION_GENERAL_MIN              "1.0.0"
+#define ADDON_GLOBAL_VERSION_GENERAL_XML_ID           "kodi.binary.global.general"
+#define ADDON_GLOBAL_VERSION_GENERAL_DEPENDS          "General.h"
```
There coming on next requests two another global parts, one for network and one for filesystem. Then is about globals for the moment everything done.

Will keep them apart to avoid possible problems with changes.

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  }
+}
+
+void Interface_General::open_settings_dialog(void* kodiBase)
```
Shouldn't this function have a bool return value indicating success as there are several error cases?

AlwinEsch replied:
I look what possible

AlwinEsch replied:
Was a bit a c&p from Python addon interface, becomes changed :-D

AlwinEsch commented:
Aside from my bad English, is it OK to bring this?

Because of the English, wants in the end when everything is there in the doxygen things make and with the help of you make the Doku fit. Here how I planned the docu: http://alwinesch.github.io/group__cpp.html

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  return buffer;
+}
+
+void Interface_General::get_dvd_menu_language(void* kodiBase, char& language, unsigned int& iMaxStringSize)
```
For consistency reasons (see get_localized_string) I would let this function return a char* containing the language. Another benefit is that you do not need iMaxStringSize any longer.

AlwinEsch replied:
Becomes changed.

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  iMaxStringSize = str.length();
+}
+
+void Interface_General::queue_notification(void* kodiBase, const int type, const char* message)
```
boolean return value - here and for every function currently returning void?

AlwinEsch replied:
Is for a KaiToast Message who have already "void" as return, no need to have a return for me.

MartijnKaijser commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+
+void Interface_General::DeInit(AddonGlobalInterface* addonInterface)
+{
+  if (addonInterface->toKodi && /* <-- Safe check, needed so long old addon way is present */
```
needed as long as the old addon way is used

MartijnKaijser commented on xbmc/addons/interfaces/General.h:
```diff
@@ -0,0 +1,65 @@
+    /*!
+     * @brief callback functions from add-on to kodi
+     *
+     * @note For add of new functions use the "_" style to identify direct a
```
"To add a new function use the "_" style to directly identify an add-on callback function. Everything with CamelCase is only to be used in Kodi."

ksooo reviewed (commented):
A bunch of questions, suggestions for improvement...

MartijnKaijser commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h:
```diff
@@ -0,0 +1,385 @@
+ * All function pointers there are used by the C++ interface functions below.
+ * You find the set of them on xbmc/addons/interfaces/General.cpp
+ *
+ * Note: For add-on development itself not needed, thats why with '*' here in
```
"add-on development itself this is not needed"

MartijnKaijser commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h:
```diff
@@ -0,0 +1,385 @@
+/*
+ * For interface between add-on and kodi.
+ *
+ * In this structure becomes the addresses of functions inside Kodi stored who
```
"This structure defines the addresses of functions stored inside Kodi which are then available for the add-on to call"

MartijnKaijser commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h:
```diff
@@ -0,0 +1,385 @@
+/// @note Label id's \b 30000 to \b 30999 and \b 32000 to \b 32999 are related
+/// to own add-on included strings inside
+/// <b>./resources/language/resource.language.??_??/strings.po</b>
+/// all others are from Kodi itself.
```
" are related to the add-on's own included strings from ./resources/language/resource.language.??_??."
"All other strings are from Kodi core language files."

MartijnKaijser commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h:
```diff
@@ -0,0 +1,385 @@
+/// \ingroup cpp_kodi
+/// @brief Returns an addon's localized 'unicode string'.
+///
+/// @param[in] labelId    for string you want to localize.
```
"string you want to localize"

MartijnKaijser commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h:
```diff
@@ -0,0 +1,385 @@
+///
+/// @param[in]  stringSrc       The string to translate.
+/// @param[out] utf8StringDst   The translated string.
+/// @param[in]  failOnBadChar   [opt] return failed if bad character is inside <em>(default is <b><c>false</c></b>)</em>
```
.." return**s** failed"

MartijnKaijser reviewed (commented):
Did a quick English check with some improvement points. Perhaps @jjd-uk or any other native English speaker can do a check. We can always fix up later of course.

AlwinEsch commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  addonInterface->toKodi->kodi->queue_notification = queue_notification;
+  addonInterface->toKodi->kodi->queue_notification_from_type = queue_notification_from_type;
+  addonInterface->toKodi->kodi->queue_notification_with_image = queue_notification_with_image;
+  addonInterface->toKodi->kodi->get_dvd_menu_language = get_dvd_menu_language;
```
Currently use a addon a bad way to identify the used language ;-) from Kodi, but can also be global for me.

AlwinEsch commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  return buffer;
+}
+
+char* Interface_General::unknown_to_utf8(void* kodiBase, const char* source, bool& ret, bool failOnBadChar)
```
You mean the "bool&"?

AlwinEsch replied:
Becomed changed to bool*

AlwinEsch commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  else
+    string = g_localizeStrings.Get(dwCode).c_str();
+
+  char* buffer = strdup(string.c_str());
```
Free becomes done here https://github.com/AlwinEsch/kodi/blob/0290f7a998308b21e224571315f7e252a06e2b04/xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h#L124

ksooo replied:
Ah, okay.

AlwinEsch commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+  addonInterface->toKodi->kodi->unknown_to_utf8 = unknown_to_utf8;
+  addonInterface->toKodi->kodi->queue_notification = queue_notification;
+  addonInterface->toKodi->kodi->queue_notification_from_type = queue_notification_from_type;
+  addonInterface->toKodi->kodi->queue_notification_with_image = queue_notification_with_image;
```
On addon itself you have only "QueueNotification" but for the "C" must be another name for the types.

With the different ways of Kodi becomes it hard to hold easy for addons, hope the docu can help addon developers http://alwinesch.github.io/group__cpp__kodi.html#gacb09570070a6a601974ecb76ab9e521b

AlwinEsch commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,198 @@
+    return nullptr;
+
+  std::string string;
+  if ((dwCode >= 30000 && dwCode <= 30999) || (dwCode >= 32000 && dwCode <= 32999))
```
Is a remnant from the Python interface, hope to get improved in the future

ksooo replied:
Okay.

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,202 @@
+  AddonPtr addonInfo;
+  if (CAddonMgr::GetInstance().GetAddon(addon->ID(), addonInfo))
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - Could not get addon information for '%s'", __FUNCTION__, addon->ID().c_str());
```
<code>return false;</code> missing here!

AlwinEsch commented:
Have cleaned up and improved the commented places.

Thanks a lot!

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,202 @@
+  return buffer;
+}
+
+char* Interface_General::get_dvd_menu_language(void* kodiBase)
```
You said that this function is "misused" by addons to obtain Kodi UI language and this is the reason for putting this function to global.general. I suggest to rename (and reimplement) this function to reflect what the addon devs actually need => get_ui_language maybe? 

get_dvd_menu_language should be put into those addon types where dvd menu language actually makes  sense.

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,202 @@
+
+void Interface_General::queue_notification_with_image(void* kodiBase, const char* imageFile, const char* caption,
+                                                      const char* description, unsigned int displayTime,
+                                                      bool withSound, unsigned int messageTime)
```
I still don't get why we have three different queue_notification functions. All three function signatures are very similar. 

IMO, one function is sufficient:

<pre>
bool Interface_General::queue_notification(void* kodiBase, int type, const char* imageFile, const char* caption,
                                                      const char* description, unsigned int displayTime,
                                                      bool withSound, unsigned int messageTime)
</pre>

This covers all cases.

The boolean return value is useful for covering error cases (invalid paramter combinations, values, ...).

ksooo reviewed (changes requested):
Still some questions, suggestions...

AlwinEsch committed:
[addons] improve GetAddonPath and GetBaseUserPath callbacks

With this change becomes it allowed to append a string to returned
path.

This can reduce code lines on addon.

AlwinEsch commented:
Is updated and `QueueNotification` changed to only one function.

ksooo commented on xbmc/addons/interfaces/General.cpp:
```diff
@@ -0,0 +1,226 @@
+  return buffer;
+}
+
+bool Interface_General::queue_notification(void* kodiBase, const int type, const char* header,
```
<code>const int</code> makes no sense, just <code>int</code>, please.

ksooo reviewed (commented):
Just one minor nitpick left. Thanks for being patient with me. ;-)

AlwinEsch committed:
[addons] add general addon callback functions to new style

This add the general function who used on several addons to the new addon
interface style.

In future all the basics like on Python addons can be added here.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
@ksooo no problem, you welcome and you help me a lot!

ksooo commented:
Add far as I can see - ready to merger now...
